src/ctap/hid/mod.rs

```diff
@@ -713,35 +639,188 @@ mod test {
             &mut ctap_state,
-            vec![Message {
-                cid,
-                cmd: CtapHidCommand::Ping,
-                payload: vec![0x99, 0x99],
-            }],
         );
+        assert_eq!(response.next(), Some(ping_packet));
+        assert_eq!(response.next(), None);
+    }
+
+    #[test]
+    fn test_process_hid_packet_empty() {
+        let mut env = TestEnv::new();
+        let mut ctap_state = CtapState::new(&mut env, CtapInstant::new(0));
+        let (mut ctap_hid, cid) = CtapHid::new_initialized();
+
+        let mut cancel_packet = [0x00; 64];
+        cancel_packet[..4].copy_from_slice(&cid);
+        cancel_packet[4..7].copy_from_slice(&[0x91, 0x00, 0x00]);
+
+        let mut response = ctap_hid.process_hid_packet(
+            &mut env,
+            &cancel_packet,
+            CtapInstant::new(0),
+            &mut ctap_state,
+        );
+        assert_eq!(response.next(), None);
+    }
+
+    #[test]
+    fn test_wink() {
+        let mut env = TestEnv::new();
+        let mut ctap_state = CtapState::new(&mut env, CtapInstant::new(0));
+        let (mut ctap_hid, cid) = CtapHid::new_initialized();
+        assert!(!ctap_hid.should_wink(CtapInstant::new(0)));
+
+        let mut wink_packet = [0x00; 64];
+        wink_packet[..4].copy_from_slice(&cid);
+        wink_packet[4..7].copy_from_slice(&[0x88, 0x00, 0x00]);
 
+        let mut response = ctap_hid.process_hid_packet(
+            &mut env,
+            &wink_packet,
+            CtapInstant::new(0),
+            &mut ctap_state,
+        );
+        assert_eq!(response.next(), Some(wink_packet));
+        assert_eq!(response.next(), None);
+        assert!(ctap_hid.should_wink(CtapInstant::new(0)));
+    }
+
+    #[test]
+    fn test_split_message() {
+        let message = Message {
+            cid: [0x12, 0x34, 0x56, 0x78],
+            cmd: CtapHidCommand::Ping,
+            payload: vec![0x99, 0x99],
+        };
+        let mut response = CtapHid::split_message(message);
+        let mut expected_packet = [0x00; 64];
+        expected_packet[..9]
+            .copy_from_slice(&[0x12, 0x34, 0x56, 0x78, 0x81, 0x00, 0x02, 0x99, 0x99]);
+        assert_eq!(response.next(), Some(expected_packet));
+        assert_eq!(response.next(), None);
+    }
+
+    #[test]
+    fn test_split_message_too_large() {
+        let payload = vec![0xFF; 7609 + 1];
+        let message = Message {
+            cid: [0x12, 0x34, 0x56, 0x78],
+            cmd: CtapHidCommand::Cbor,
+            payload,
+        };
+        let mut response = CtapHid::split_message(message);
+        let mut expected_packet = [0x00; 64];
+        expected_packet[..8].copy_from_slice(&[0x12, 0x34, 0x56, 0x78, 0x90, 0x00, 0x01, 0xF2]);
+        assert_eq!(response.next(), Some(expected_packet));
+        assert_eq!(response.next(), None);
+    }
+
+    #[test]
+    fn test_keepalive() {
+        for status in [KeepaliveStatus::Processing, KeepaliveStatus::UpNeeded].iter() {
```

Comment:
nit/fyi: it's possible to write `&status` here (so once) instead of `*status` at each usage occurrences below (when the type is copy).

Reply:
Better, fixed.